Extend build system to better support package build systems

fingolfin opened the pull request:
== UPDATE

OK, now that this is rebased and the CI stuff is out, I'll try to summarize the changes, in the hopes that this'll make review a bit easier. They actually IMHO don't add up to that much, but if merged, they'll allow me to tackle rewriting `autoconf` based package build system once GAP 4.9 is out, w/o having to worry about backwards compatibility too much (as long as the maintainers of those packages are OK with requiring GAP 4.9). Indeed, one might even consider changing several packages using autoconf+automake now to go back to a `gac` based build system (which @ChrisJefferson might enjoy ;-), or at least a much simpler custom build system.

* Small improvements to `README.buildsys.md`
* the build system tracks the values of CFLAGS, CPPFLAGS, LDFLAGS, LIBS and OBJS. If any of them change, then appropriate actions are taken (e.g. if CFLAGS or CPPFFLAGS change, recompile all .c files; if LIBS, OBJS or LDFLAGS change, re-link). This is meant to allow things like `make -j8 CFLAGS=-O3` to quickly rebuild GAP with different settings w/o having to re-run `configure`; if you then later do `make` again, it'll switch back to the default (via `configure`) build settings.
* generate `sysinfo.gap` from the makefiles, not `configure`. This ensures it stays in think with any build system changes; it also means that when working on the build system, one can quickly modify what's in `sysinfo.gap` without having to re-run `autoconf && configure` each time, which is a major boon to me
* changed `GAPARCH` to not include the compiler; compilers these days should produce compatible binaries anyway, and this way, if you use multiple compilers (as I do) for testing, you don't have to recompile every package for every compiler (or use lots of symlinks). On the on run, I want to either get rid of `GAPARCH` resp. refine it massively, but that's not yet here; still, this small change is a first step in that direction
* Rewrote `gac` to work well with the new build system, and perhaps better than before.
* It is again possible to use `gac` to generate "static" modules, i.e. it produces a new GAP executable with your module statically linked in. I am not sure that's so useful these days, and we might want to drop this eventually; but in the meantime, it was a fun exercise. I also extended the test suite to cover this.
* More information is stored in `sysinfo.gap`, which is used by `gac`, and which I plan to use in package build systems in general. In the future, we may want to store even more information about the build in it. But for now, it already is enough to simplify some bits of package build systems: Namely figuring out the correct include flags to access GAP headers, and also suitable linker flags, and even which C compiler was use. This can be used to write custom package build systems which involve neither `autoconf` and `automake`, nor `gac` (and even if `autoconf` is used, it'll simplify some things).
 A `sysinfo.gap` now looks like this (here for a 64bit HPC-GAP out-of-tree build):
```
# This files has been generated by the GAP build system,
# do not edit manually!
GAParch=x86_64-apple-darwin15.6.0-hpcgap64
GAP_ABI=64

GAP_BIN_DIR="/Users/mhorn/Projekte/GAP/gap.github/build/hpcgap"
GAP_LIB_DIR="/Users/mhorn/Projekte/GAP/gap.github/build/hpcgap/../.."

GAP_CC="gcc"
GAP_CFLAGS=" -Qunused-arguments -pthread -g -O2 "
GAP_CPPFLAGS=" -I/Users/mhorn/Projekte/GAP/gap.github/build/hpcgap -I/Users/mhorn/Projekte/GAP/gap.github/build/hpcgap/../.. -DHAVE_CONFIG_H   -I/Users/mhorn/Projekte/GAP/gap.github/build/hpcgap/extern/install/gc/include -I/Users/mhorn/Projekte/GAP/gap.github/build/hpcgap/extern/install/libatomic_ops/include -isystem /sw/include"
GAP_LDFLAGS="   /Users/mhorn/Projekte/GAP/gap.github/build/hpcgap/extern/install/gc/lib/libgc.la /Users/mhorn/Projekte/GAP/gap.github/build/hpcgap/extern/install/libatomic_ops/lib/libatomic_ops.la -Qunused-arguments -pthread  -L/sw/lib "
GAP_LIBS=" -lgmp -lreadline   "

GAP_OBJS="obj/ariths.lo obj/blister.lo obj/bool.lo obj/c_methsel1.lo obj/c_oper1.lo obj/c_type1.lo obj/calls.lo obj/code.lo obj/compiler.lo obj/compstat.lo obj/costab.lo obj/cyclotom.lo obj/debug.lo obj/dt.lo obj/dteval.lo obj/exprs.lo obj/ffdata.lo obj/finfield.lo obj/funcs.lo obj/gap.lo obj/gap_version.lo obj/gapstate.lo obj/gvars.lo obj/hookintrprtr.lo obj/integer.lo obj/intfuncs.lo obj/intrprtr.lo obj/io.lo obj/iostream.lo obj/listfunc.lo obj/listoper.lo obj/lists.lo obj/macfloat.lo obj/objccoll.lo obj/objcftl.lo obj/objects.lo obj/objfgelm.lo obj/objpcgel.lo obj/objscoll.lo obj/objset.lo obj/opers.lo obj/permutat.lo obj/plist.lo obj/pperm.lo obj/precord.lo obj/profile.lo obj/range.lo obj/rational.lo obj/read.lo obj/records.lo obj/saveload.lo obj/scanner.lo obj/sctable.lo obj/set.lo obj/stats.lo obj/streams.lo obj/stringobj.lo obj/sysfiles.lo obj/system.lo obj/tietze.lo obj/trans.lo obj/vars.lo obj/vec8bit.lo obj/vecffe.lo obj/vecgf2.lo obj/vector.lo obj/weakptr.lo obj/boehm_gc.lo obj/hpc/aobjects.lo obj/hpc/misc.lo obj/hpc/serialize.lo obj/hpc/thread.lo obj/hpc/threadapi.lo obj/hpc/tls.lo obj/hpc/traverse.lo"
```


== Things I'd like to see but which are not yet in there

... include ...
* C++ support in `gac`
* exposing info about `libtool` via 
* `make install` support (but I expect `sysinfo.gap` to play a crucial role in that, and for that it'll be super helpful to be able to quickly change what the build system puts in there)
* a way to share a single package directory among multiple GAP builds (possibly out-of-tree builds) in a convenient way. 



== old PR description follows

This is work in progress on improving our build system, with the goal of 
* ... fixing gac for various uses cases (and possibly removing use cases we don't want to support anymore),
* making it easier for packages to support out-of-tree builds,
* making it easier for packages to support HPC-GAP builds,
* other tweaks.

This is WIP, and not yet cleaned up; some things are still in flux. Things to do include:

1. either fix `gac` support for "static" loadable modules, or remove it
2. add tests for various `gac` modes (may be postponed till after GAP 4.9, though)
3. figure out how package build systems should interact with multiple out-of-tree builds suitably (I have several ideas on how to do that.
4. ... more ... need to look through my notes ...
5. adapt a few packages to make use of the changes in this PR

Right now, the last point is most important. Ideally, the changes will be so that one can still use the packages with GAP 4.8.8. But that may turn out to be a lot of work, so it would be way simpler if I could just make the changes so that packages will require GAP 4.9 (resp. master branch). Thoughts?

(Oh yeah: *old* packages should still be working with this PR, which is part of the reason I am submitting it now: so that I get travis feedback :-).

alex-konovalov commented:
From `make testpackagesload` on Linux 64-bit I see only packages fail to load are: anupq, fwtree, itc, linboxing, pargap, polymakeinterface, xgap. But due to dependencies, the real fails are anupq and polymakeinterface only. But anupq fails because of not finding GMP 
```
...
  CC       src/pq-write.o
  CC       src/pq-main.o
  CCLD     pq
/usr/bin/ld: cannot find -lgmp
src/pq-TemporaryFile.o: In function `TemporaryFile':
TemporaryFile.c:(.text+0x15): warning: the use of `tempnam' is dangerous, better use `mkstemp'
collect2: error: ld returned 1 exit status
gmake: *** [pq] Error 1

WARNING: Failed to build anupq-3.1.5
```
and polymakeinterface fails because polymake is not installed on that machine:
```
==== Checking PolymakeInterface
Running './configure' '/circa/scratch/gap-jenkins/workspace/GAP-pull-request-quicktest/GAPCOPTS/64build/GAPGMP/gmp/GAPTARGET/packagesload/label/kovacs/GAP-pull-request-snapshot' 
./configure: line 23: polymake-config: command not found
./configure: line 25: polymake-config: command not found
./configure: line 27: polymake-config: command not found
./configure: line 29: polymake-config: command not found
Running 'gmake' 
mkdir -p bin/x86_64-pc-linux-gnu-default64
cp /circa/scratch/gap-jenkins/workspace/GAP-pull-request-quicktest/GAPCOPTS/64build/GAPGMP/gmp/GAPTARGET/packagesload/label/kovacs/GAP-pull-request-snapshot/bin/x86_64-pc-linux-gnu-default64/config.h bin/x86_64-pc-linux-gnu-default64/
g++  -fPIC -Wall -O0 -g -pipe -m64 -o bin/x86_64-pc-linux-gnu-default64/polymake_fktn.o -I/circa/scratch/gap-jenkins/workspace/GAP-pull-request-quicktest/GAPCOPTS/64build/GAPGMP/gmp/GAPTARGET/packagesload/label/kovacs/GAP-pull-request-snapshot/bin/x86_64-pc-linux-gnu-default64/extern/gmp/include -I/circa/scratch/gap-jenkins/workspace/GAP-pull-request-quicktest/GAPCOPTS/64build/GAPGMP/gmp/GAPTARGET/packagesload/label/kovacs/GAP-pull-request-snapshot/bin/x86_64-pc-linux-gnu-default64/../.. -I/circa/scratch/gap-jenkins/workspace/GAP-pull-request-quicktest/GAPCOPTS/64build/GAPGMP/gmp/GAPTARGET/packagesload/label/kovacs/GAP-pull-request-snapshot/bin/x86_64-pc-linux-gnu-default64 -DCONFIG_H -c src/polymake_fktn.cpp	
In file included from src/polymake_fktn.h:2:0,
from src/polymake_fktn.cpp:1:
src/polymake_data.h:6:27: fatal error: polymake/Main.h: No such file or directory
#include <polymake/Main.h>
^
compilation terminated.
gmake: *** [bin/x86_64-pc-linux-gnu-default64/polymake_fktn.o] Error 1

WARNING: Failed to build PolymakeInterface
```

fingolfin commented:
@alex-konovalov it's not quite clear to me what you are saying here: are these failures that occur with the branch in this PR, but *not* with the master branch?!?

alex-konovalov commented:
@fingolfin sorry, was unclearly formulated in a rush. I meant that the two failures in anupq and PolymakeInterface are not caused by your PR since they have another explanation. And yes, same observed in the master branch.

I've no started next Jenkins jib, which will exercise building packages on Linux and macOS in both 32- and 64-bit builds.

alex-konovalov commented:
I've tested this PR again (with "Refactor CI setup for Travis and AppVeyor" https://github.com/gap-system/gap/pull/1906/commits/875a8c98b2340d4fa490a4415259616b71df96d5 commit being the latest) and from Jenkins PoV it looks good - no regressions with building packages are detected.

fingolfin commented:
I have now also fixed static compilation with gac, and added tests for that.

It would be really good to get this merged into GAP 4.9, because then once GAP 4.9 is out, we can start converting packages to take advantage of the new content of `sysinfo.gap`.

ChrisJefferson commented on cnf/gac.in:
```diff
@@ -63,33 +63,37 @@ gap_bin="@abs_top_builddir@"
+# read sysinfo.gap, which should set GAP_CFLAGS, GAP_CPPFLAGS, GAP_LDFLAGS, GAP_LIBS
+. "${abs_top_builddir}/sysinfo.gap"
+
+echo "gac HACK: GAP_CPPFLAGS = ${GAP_CPPFLAGS}"
```
This be cleaned up?

fingolfin replied:
oops, of course, thanks

alex-konovalov commented:
Re-checked with Jenkins last night - no regressions detected. Did not look at actual diffs though.

fingolfin commented:
I moved most of the Travis/CI related changes to a new PR #1979. There is still some overlap between the two, though, for the static gac tests. If one of the two PRs gets merged, I can rebase the other.

fingolfin commented:
FYI, I update the description text.

ChrisJefferson commented:
I am entirely happy with this, but I might let @alex-konovalov merge it

stevelinton commented:
Happy with this, although I haven't reviewed it line-by-line
